### ISSUE TITLE
4.9 regression on conditional type

### ISSUE THREAD

**Issue by testerez** [microsoft/TypeScript#51540]
# Bug Report

### 🔎 Search Terms

4.9 Type 'number' does not satisfy the constraint

### 🕗 Version & Regression Information

This changed between versions 4.8 and 4.9.3

### ⏯ Playground Link

[tsplay.dev/NdrGyW](https://tsplay.dev/NdrGyW)

### 💻 Code

```ts repro
type Decrement<N extends number> = [-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10][N]

type PathImpl<T, K extends keyof T, Depth extends number> = Depth extends 0
  ? ''
  : K extends string | number
  ? T[K] extends Record<string, any>
    ? T[K] extends ArrayLike<any>
      ? K | `${K}.${PathImpl<T[K], number, Decrement<Depth>>}`
      : K | `${K}.${PathImpl<T[K], keyof T[K], Decrement<Depth>>}`
    : K
  : never

export type ObjectPath<T> = PathImpl<T, keyof T, 10> | keyof T

```

### 🙁 Actual behavior

```
Type 'number' does not satisfy the constraint 'keyof T[K]'.
```

### 🙂 Expected behavior

Since `T[K] extends ArrayLike<any>` then I should be able to index by number. This was working on 4.8

**Comment by Tobeyyyyy** [comment]
Here is a slightly smaller example to reproduce the issue.

[Playground](https://www.typescriptlang.org/play?ts=5.0.0-dev.20221115#code/FAFwngDgpgBACgQxACwJIFsIBsA8AVAGhgGkYoAPEKAOwBMBnGAayjAHsAzGPAPhgF4YwGCTKUaDGPRAAnAJbUA5jAA+MagFd0AIygzhMAPzcA2sQC6YqnUYAlKAGM2M2jmnylRBNTA8DI4zwzSwprSQBBGRkEMAAZORYcb19-ESN4JDRMXCCLIk0dPT80tIAudSgANz1U8uoqmpE6hv0gA)

```
type PathImpl<T, K extends keyof T> = 
  K extends string | number
  ? T[K] extends Record<string, any>
    ? T[K] extends ArrayLike<any>
      ? PathImpl<T[K], number>
      : never
    : never
  : never
```

All three conditionals are needed to reproduce the error.

**Comment by cjbarth** [comment]
I think we might have a similar issue, but it is related to `keyof typeof this.privateMember`. See https://github.com/node-saml/node-saml/issues/218

**Comment by JGJP** [comment]
Seeing a similar issue in `@storybook/components` popperjs typing, it fails with this message:

```
Type 'Options' does not satisfy the constraint 'Obj'
```

where the code looks like this:

```tsx
export declare type Modifier<Name, Options> = {
    name: Name;
    enabled: boolean;
    phase: ModifierPhases;
    requires?: Array<string>;
    requiresIfExists?: Array<string>;
    fn: (arg0: ModifierArguments<Options>) => State | void; // <-- fails here
    effect?: (arg0: ModifierArguments<Options>) => (() => void) | void; // <-- and here
    options?: Partial<Options>;
    data?: Obj;
};

export declare type ModifierArguments<Options extends Obj> = {
    state: State;
    instance: Instance;
    options: Partial<Options>;
    name: string;
};

export declare type Obj = {
    [key: string]: any;
};
```

So it looks like it's failing to assign an untyped generic arg to `{ [key: string]: any }`. Is this expected?

**Comment by RyanCavanaugh** [comment]
@JGJP that looks like a correct error to me (kinda hard to say since the example is self-contained). `Options` is unconstrained; if you believe that an unconstrained type parameter should be allowable to pass to a type parameter constrained to `Obj`, why did you write the `extends Obj` constraint in the first place?

**Comment by JGJP** [comment]
@RyanCavanaugh this is not my code, it's from popperjs typings, used by `@storybook/components`. It's only broken after upgrading TS, so I assume it was passing type checks before (4.6)

**Comment by ahejlsberg** [comment]
This issue no longer reproduces. Looks like it was fixed in TS 5.0.
